src/genparticles.cxx

```diff
@@ -0,0 +1,305 @@
+    return df1;
+}
+/**
+ * @brief function to find all hadronicGenTaus needed for the genmatching, based
+ * on
+ * https://github.com/KIT-CMS/Artus/blob/dictchanges/KappaAnalysis/src/Utility/GeneratorInfo.cc
+ * implementation. Loop trough all genparticles and check, if a genparticle,
+ that is prompt, without any leptonic daughters can be found. If this
+ genparticle has a neutrino as daughter, the hadronicGenTau is added to the list
+ of hadronicGenTaus.
+ *
+ * @param df The dataframe to be extended with the hadronicGenTaus
+ * @param outputname The name of the column to be added to the dataframe
+ * @param genparticles_pdgid The name of the column containing the pdgids of the
+ genparticles
+ * @param genparticles_statusFlag The name of the column containing the
+ statusflags of the genparticles
+ * @param genparticles_motherid The name of the column containing the motherids
+ of the genparticles
+ * @return a new dataframe with the hadronicGenTaus added to the dataframe
+
+ */
+
+ROOT::RDF::RNode hadronicGenTaus(ROOT::RDF::RNode df,
+                                 const std::string &outputname,
+                                 const std::string &genparticles_pdgid,
+                                 const std::string &genparticles_statusFlag,
+                                 const std::string &genparticles_motherid) {
+
+    auto gentaus = [](const ROOT::RVec<int> &pdgids,
+                      const ROOT::RVec<int> &status_flags,
+                      const ROOT::RVec<int> &mother_index) {
+        // set default values for the output
+        std::vector<int> hadronicGenTaus;
+        if (pdgids.size() == 0) {
+            hadronicGenTaus.push_back(-1);
+            hadronicGenTaus.push_back(-1);
+            return hadronicGenTaus;
+        }
+        // debug printout
+        for (int i = 0; i < pdgids.size(); i++) {
+            Logger::get("genmatching::tau::genpair")
+                ->debug("genparticle index {}, pdgid: {}, status_flags {}, "
+                        "mother_index {}",
+                        i, pdgids.at(i), status_flags.at(i),
+                        mother_index.at(i));
+        }
+        // now loop though the genparticles and find the taus
+        for (unsigned int i = 0; i < pdgids.size(); i++) {
+            // check if the particle is a tau
+            if (std::abs(pdgids.at(i)) == 15) {
+                // check if the particle is a stable one
+                bool prompt = IntBits(status_flags.at(i)).test(0);
+                if (prompt) {
+                    Logger::get("genmatching::tau::genpair")
+                        ->debug("Found prompt tau: {}", i);
+                    // find all daughters of the tau by checking which particles
+                    // have the tauindex i as mother
+                    std::vector<int> daughters;
+                    for (unsigned int j = 0; j < mother_index.size(); j++) {
+                        if (mother_index.at(j) == i) {
```

Comment:
This only checks for direct daughters, is this sufficient?

Reply:
jip, should be sufficient